bqskit/passes/synthesis/diagonal.py

```diff
@@ -0,0 +1,108 @@
+"""This module implements the DiagonalSynthesisPass."""
+from __future__ import annotations
+
+import logging
+
+from numpy import where
+
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.gates import CNOTGate
+from bqskit.ir.gates import RZGate
+from bqskit.passes.synthesis.synthesis import SynthesisPass
+from bqskit.qis.state.state import StateVector
+from bqskit.qis.state.system import StateSystem
+from bqskit.qis.unitary import UnitaryMatrix
+from bqskit.utils.math import pauliz_expansion
+from bqskit.utils.math import unitary_log_no_i
+
+
+_logger = logging.getLogger(__name__)
+
+
+class DiagonalSynthesisPass(SynthesisPass):
+    """A pass that synthesizes diagonal unitaries."""
+
+    def __init__(
+        self,
+        parameter_precision: float = 1e-8,
+    ) -> None:
+        """
+        Constructor for DiagonalSynthesisPass.
+
+        Args:
+            parameter_precision (float): Pauli strings with parameter values
+                less than this are rounded to zero. (Default: 1e-8)
+
+        TODO:
+            - Optimize Pauli string ordering
+            - Cancel adjacent CNOTs
+        """
+        self.parameter_precision = parameter_precision
+
+    def gray_code(self, number: int) -> int:
+        """Convert a number to its Gray code representation."""
+        gray = number ^ (number >> 1)
+        return gray
+
+    def pauli_to_subcircuit(
+        self,
+        string_id: int,
+        angle: float,
+        num_qubits: int,
+    ) -> Circuit:
+        string = bin(string_id)[2:].zfill(num_qubits)
+        circuit = Circuit(num_qubits)
+        locations = [i for i in range(num_qubits) if string[i] == '1']
+        if len(locations) == 1:
+            circuit.append_gate(RZGate(), locations[0], [angle])
+        elif len(locations) > 1:
+            pairs = [
+                (locations[i], locations[i + 1])
+                for i in range(len(locations) - 1)
+            ]
+            for pair in pairs:
+                circuit.append_gate(CNOTGate(), pair)
+            circuit.append_gate(RZGate(), locations[-1], [angle])
+            for pair in reversed(pairs):
+                circuit.append_gate(CNOTGate(), pair)
+        return circuit
+
+    async def synthesize(
+        self,
+        utry: UnitaryMatrix | StateVector | StateSystem,
+        data: PassData,
+    ) -> Circuit:
+        """Synthesize `utry`, see :class:`SynthesisPass` for more."""
+        if not isinstance(utry, UnitaryMatrix):
+            m = 'DiagonalSynthesisPass can only synthesize diagonal, '
+            m += f'UnitaryMatrixs, got {type(utry)}.'
```

Comment:
The plural UnitaryMatrixs is bothering me. A nit pick for sure. I know they are object names and as such you just want to add an s to the end of it, but can we surround it with backticks or quotes? "`UnitaryMatrix`s"

Your call